### ISSUE TITLE
Webhook service randomly timeouts

### ISSUE THREAD

**Issue by fierman333** [clastix/capsule#439]
# Bug description

Webhook service is randomly denying https calls:
```
Internal error occurred: failed calling webhook "namespaces.capsule.clastix.io": Post "https://capsule-webhook-service.capsule-system.svc:443/namespaces?timeout=30s": dial tcp 100.69.203.179:443: connect: connection refused
```

if I try manually to call that service from another pod for example I am getting randomly `Connection refused` error:
```
| => kubectl exec somepod  -- curl -I -k https://capsule-webhook-service.capsule-system.svc:443/namespaces?timeout=30
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0   128    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
HTTP/1.1 200 OK
Date: Tue, 28 Sep 2021 08:51:59 GMT
Content-Length: 128
Content-Type: text/plain; charset=utf-8

| => kubectl exec somepod -- curl -I -k https://capsule-webhook-service.capsule-system.svc:443/namespaces?timeout=30
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (7) Failed to connect to capsule-webhook-service.capsule-system.svc port 443 after 3 ms: Connection refused
command terminated with exit code 7
```

Something which we are noticing that if we run `capsule-controller-manager` deployment with one pod then we do not have that timeouts, but if we run it in '3' pods then seems that only 1 of 3 pods response on https calls when we test webhook service via port-forward:
```
kubectl -n capsule-system port-forward capsule-controller-manager-d6f6b45d-4h9b2 9443
Forwarding from 127.0.0.1:9443 -> 9443
Forwarding from [::1]:9443 -> 9443
```

```
curl -k https://localhost:9443/namespaces
curl: (35) LibreSSL SSL_connect: SSL_ERROR_SYSCALL in connection to localhost:9443
```

If we run above command on the pod which returns response then we get:
```
curl -k https://localhost:9443/namespaces
{"response":{"uid":"","allowed":false,"status":{"metadata":{},"message":"contentType=, expected application/json","code":400}}}
```

Webhook service should work properly on all pods in the deployment right ?
Otherwise this would mean that `capsule-controller-manager` deployment should run with 1 pod.

# How to reproduce

Run some `kubectl` command as `kubectl label ${namespace} xxx=xxx` for example

1. No Capsule Tenant YAML defined

# Expected behavior

Executing `kubectl` commands should not return capsule webhook error

# Logs

No errors in `capsule-controller-manager` pods

# Additional context

- Capsule version: 0.1.0
- Helm Chart version: capsule-0.1.0
- Kubernetes version: v1.19.12

**Comment by prometherion** [comment]
Hey @fierman333, thanks for this bug report, definitely something we have to look at! 👀 

Prior to debugging, would be great having more context, such as the resources assigned to the Capsule Manager pods, if you ever notice something wrong from the metrics perspective, such as stale goroutines, or too high GC time, or high request on the inner REST API client.

If you had the chance to collect these metrics, would be great having some data to try debugging it in a remote way, in the meanwhile I could try replicating this in one of our labs.

I know @MaxFedotov is running Capsule in several production environments, wondering if he's facing the same issue or not, and which setup is using (a single or multiple replicas).

**Comment by prometherion** [comment]
JFI I was trying to replicate unluckily this issue so far, sharing my setup.

1. creating a Kubernetes cluster with Capsule installed
2. waiting for certs, etc.
3. scaling with `kubectl -n capsule-system scale deployment capsule-controller-manager --replicas=3`
4. installing a dummy Tenant manifest, such as `kubectl apply -f config/samples/capsule_v1beta1_tenant.yaml`
5. in another window, `kubectl -n capsule-sytem -fl app.kubernetes.io/name=capsule --prefix=true | grep webhook | grep -Po '^\[.(.*)]' | sort | uniq -c`
6. in another window, `export SECONDS=60; export secs=0; while (( SECONDS < secs )); do kubectl replace -f config/samples/capsule_v1beta1_tenant.yaml; done`
7. once finished, killing the `kubectl logs` watch to let the pipe elaborate the result, as follows.

```
    860 [pod/capsule-controller-manager-7b5f77799c-cwldj/manager]
      4 [pod/capsule-controller-manager-7b5f77799c-dt9pq/manager]
      1 [pod/capsule-controller-manager-7b5f77799c-hlgrj/manager]
```

This test wanted to check if the replica setup is able to provide high availability on webhooks, and it seems so, also considering that the `while` loop performing `kubectl replace` was interacting with the API Server doing a `PUT` request to replace the entire Tenant manifest, which requires interaction with one of the Capsule webhooks available.

We could start discussing why the RR is not working as expected but I'd say it's an issue of my cluster setup since it's implemented on `kind` using its _fake_ CNI.

Would be great if @fierman333 is able to provide us with further details in order to replicate this, such as the CNI used, or any other valuable information.

Otherwise, we have to close this due to inactivity and that would be really bad, although aimed to understand what's not going on since this could happen to other users, too.

**Comment by fierman333** [comment]
Now sometimes we have that issue only with one running pod.
Usually this is fixed by deleting `capsule-controller-manager` pod.
Something which we notice is that when webhook is failing following log is missing:

```
{"level":"info","ts":"2022-01-21T09:27:54.439Z","logger":"controller-runtime.builder","msg":"skip registering a mutating webhook, admission.Defaulter interface is not implemented","GVK":"capsule.clastix.io/v1alpha1, Kind=Tenant"}
{"level":"info","ts":"2022-01-21T09:27:54.439Z","logger":"controller-runtime.builder","msg":"skip registering a validating webhook, admission.Validator interface is not implemented","GVK":"capsule.clastix.io/v1alpha1, Kind=Tenant"}
{"level":"info","ts":"2022-01-21T09:27:54.439Z","logger":"controller-runtime.webhook","msg":"registering webhook","path":"/convert"}
{"level":"info","ts":"2022-01-21T09:27:54.439Z","logger":"controller-runtime.builder","msg":"conversion webhook enabled","object":{"name":""}}
{"level":"info","ts":"2022-01-21T09:27:54.439Z","logger":"controller-runtime.webhook","msg":"registering webhook","path":"/pods"}
{"level":"info","ts":"2022-01-21T09:27:54.439Z","logger":"controller-runtime.webhook","msg":"registering webhook","path":"/namespaces"}
{"level":"info","ts":"2022-01-21T09:27:54.439Z","logger":"controller-runtime.webhook","msg":"registering webhook","path":"/ingresses"}
{"level":"info","ts":"2022-01-21T09:27:54.439Z","logger":"controller-runtime.webhook","msg":"registering webhook","path":"/persistentvolumeclaims"}
{"level":"info","ts":"2022-01-21T09:27:54.440Z","logger":"controller-runtime.webhook","msg":"registering webhook","path":"/services"}
{"level":"info","ts":"2022-01-21T09:27:54.440Z","logger":"controller-runtime.webhook","msg":"registering webhook","path":"/networkpolicies"}
{"level":"info","ts":"2022-01-21T09:27:54.440Z","logger":"controller-runtime.webhook","msg":"registering webhook","path":"/tenants"}
{"level":"info","ts":"2022-01-21T09:27:54.440Z","logger":"controller-runtime.webhook","msg":"registering webhook","path":"/namespace-owner-reference"}
{"level":"info","ts":"2022-01-21T09:27:54.440Z","logger":"controller-runtime.webhook","msg":"registering webhook","path":"/cordoning"}
{"level":"info","ts":"2022-01-21T09:27:54.459Z","logger":"setup","msg":"starting manager"}
I0121 09:27:54.459594       1 leaderelection.go:248] attempting to acquire leader lease vdp-system/42c733ea.clastix.capsule.io...
{"level":"info","ts":"2022-01-21T09:27:54.459Z","logger":"controller-runtime.webhook.webhooks","msg":"starting webhook server"}
{"level":"info","ts":"2022-01-21T09:27:54.459Z","logger":"controller-runtime.certwatcher","msg":"Updated current TLS certificate"}
{"level":"info","ts":"2022-01-21T09:27:54.460Z","logger":"controller-runtime.certwatcher","msg":"Starting certificate watcher"}
{"level":"info","ts":"2022-01-21T09:27:54.460Z","logger":"controller-runtime.webhook","msg":"serving webhook server","host":"","port":9443}
```

So probably service on port 9443 is not started.
Should liveness and readiness probe depend on that service when starting pods ?

**Comment by prometherion** [comment]
> Now sometimes we have that issue even only with one running pod.

Does it mean that at `t0` Capsule webhooks are working and then, suddenly, all webhooks are failing due to connection refused?

Do you have further insights on memory usage, open file descriptors or anything else that could be used to debug from an observability standpoint the _broken_ pod?

> Should liveness and readiness probe depend and check that service when starting pods?

Actually, we're just ensuring the Manager is correctly started using the default probes.

https://github.com/clastix/capsule/blob/be2678342491764c65a4745b415c6bedc8de849a/main.go#L125-L126

I double-checked and each Pod is able to serve webhooks although the underlying Controllers need leader election to avoid update conflicts. If you could build locally Capsule and test in your cluster would be good.

```diff
diff --git a/main.go b/main.go
index a040564..53df3ae 100644
--- a/main.go
+++ b/main.go
@@ -122,7 +122,7 @@ func main() {
                os.Exit(1)
        }
 
-       _ = manager.AddReadyzCheck("ping", healthz.Ping)
+       _ = manager.AddReadyzCheck("ping", manager.GetWebhookServer().StartedChecker())
        _ = manager.AddHealthzCheck("ping", healthz.Ping)
 
        ctx := ctrl.SetupSignalHandler()
```

The said change is providing an healthcheck for the readiness probe, removing the Pod from the Endpoint/EndpointSlice resource if not responding on the desired port (it's a TCP check, so we don't have to puzzle out TLS certificates).

If it looks good we can mark this bug report as a bug, although it's still unclear why the webhook webserver is not accepting connections after a while: we got several people using Capsule in production and never had any complaints.

Also regarding the reported CNI, it's out of the equation here since you're testing via `kubectl port-foward` and it's CRI responsibility.

Since we're using `controller-manager` code-base we have to ensure if this issue is related to Capsule code base or, rather, to the library one: in that case, we need to collect any information to replicate the issue to fill a bug report for the community.

**Comment by prometherion** [comment]
Thanks for the update @fierman333!

I think it's related to #440: due to that bug, we needed to skip setting up webhooks when the CA is missing, otherwise, we would have encountered issues with the CRD conversion webhook.

However, when one instance of the manager gains the leader election, it should delete the other Pod once CA has been generated.

https://github.com/clastix/capsule/blob/cd56eab119c92bf870b9995d7ac8760906048124/controllers/secret/tls.go#L115-L148

Could you check, please, if the log message `Capsule TLS certificates has been updated, Controller pods must be restarted to load new certificate` is present in one of the other pods?

**Comment by fierman333** [comment]
@prometherion Sorry for late answer. Last time when we had this issue we have decreased pods count from 3 to 1 for capsule manager and we didn't has issues for long time until now. Should this means that probably we should try with deployment count set again to 3 pods as probably that issue was fixed in newer version ?

**Comment by fierman333** [comment]
I think the problem is some race condition between installing secrets `capsule-tls` and `capsule-ca` by helm chart. On each helm chart deployment that files are redeployed as empty secrets and if I am correct that secrets should be updated.
If that secrets are not managed by helm shouldn't make sense that they should be completely managed by capsule app instead ?

**Comment by viveksyngh** [comment]
@prometherion @bsctl  IMO the generation of the `capsule-tls` and `capsule-ca`, we are doing this through built-in controllers in the capsule and trying to remove any other dependency required for the capsule to run. But can we make this feature somehow optional and provide another way to provide the tls cert and ca cert? For other operators, we use cert-manager to inject ca and tls certs for webhooks, we would like to use the same thing for the capsule as well. If we can make these controllers optional and certs path configurable then we can use `cert-manager`  to inject certificates.

**Comment by prometherion** [comment]
Thanks @fierman333, you're totally right and thanks for spotting this: now we should full take advantage of the 3-way strategic merge patch offered by Helm, please, review the PR related to this issue!

**Comment by prometherion** [comment]
> But can we make this feature somehow optional and provide another way to provide the tls cert and ca cert?

I think yes, @viveksyngh, also considering the work on #521: we could provide a Helm option to skip the creation of certificates, and reference the name of the ones created by cert-manager in the CapsuleConfiguration.

Could I ask you to open an issue? As usual, contributions are highly appreciated! 🤗

**Comment by fierman333** [comment]
@prometherion Thanks for your help fixing this issue !
I can confirm that I have tested that solution and secrets are not overwritten by helm on new deployment.
I have approved that MR.

**Comment by viveksyngh** [comment]
@prometherion I will open an issue and also work on implementing that.

**Comment by prometherion** [comment]
Chart has been released with version v0.1.7, thanks all for the hard work on tackling this!
